### PR TITLE
Table inspector Access controls

### LINKED ISSUE #2477: Access control in Table inspector

### PR DESCRIPTION
**Pull request by rajatvijay**
Fixes #2477 

**Technical details**
1. Added the ability to disable [FormBuilder](mathesar_ui/src/component-library/form-builder/FormBuilder.svelte)(the whole form in a single go) and the [FormElement](mathesar_ui/src/component-library/form-builder/FormElement.svelte)(single field at a time)
2. When the user does not have `canExecuteDDL` the whole `Properties` section is removed from the Table mode since that data is already available on the left of the `ActionsPane`

**When no canExecuteDDL**
Screenshot 2023-02-16 at 6 47 25 PM
Screenshot 2023-02-16 at 6 47 31 PM
Screenshot 2023-02-16 at 6 54 01 PM

**When no canEditTableRecords**
Screenshot 2023-02-16 at 6 55 29 PM

**When no canEditMetadata**
Screenshot 2023-02-16 at 7 00 08 PM
Screenshot 2023-02-16 at 6 58 47 PM

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by seancolsen — CHANGES_REQUESTED** [review]
This looks good except for one small issue:

- Editors and viewers should not see this save button.

    image

    There's a weird bug here. The save button appears when I change the cell selection from a cell within the "id" column to a cell within the "First Name" column. But then when I toggle between selecting a cell within the "First Name" column and a cell within the "Last Name" column, the Save button disappears and does not reappear.

**Comment by rajatvijay** [comment]
> This looks good except for one small issue:
> 
>     * Editors and viewers should not see this save button.
>       image
>       There's a weird bug here. The save button appears when I change the cell selection from a cell within the "id" column to a cell within the "First Name" column. But then when I toggle between selecting a cell within the "First Name" column and a cell within the "Last Name" column, the Save button disappears and does not reappear.

ad54b0479

File: mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte
Comment (pavish) [inline mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte:18]: Do not assign an empty string here. The input is not always a text input, for eg., for boolean we have a dropdown. The value here is a generic type which is accepted by DynamicInput and inferred by the underlying component which is based on the column type.

**Review by pavish — CHANGES_REQUESTED** [review]

File: mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte
Comment (rajatvijay) [reply mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte]: yes @pavish exactly that is why I did not re-assign it. I have pushed the final change now.

**Review by seancolsen — APPROVED** [review]
Looks good to me now. I'll wait for @pavish to re-review before merging.

File: mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte
Comment (pavish) [inline mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte:38]: There is a `initialInputValue` property in ProcessedColumn which should provide the value of the input when the default value is undefined. This eliminates the need for sanitizedValue.

I'm recommending a value & initialValue comparison with stringified values via JSON.stringify since this should provide the value as returned from the server. This is more versatile and works with types like JSON map, JSON list, and array types. The current check would not work for any of them.

```suggestion
  $: initialValue = column.column.default?.value ?? column.initialInputValue;
  $: value = initialValue;
  $: actionButtonsVisible = JSON.stringify(value) !== JSON.stringify(initialValue);
```

**Review by pavish — CHANGES_REQUESTED** [review]
@rajatvijay @seancolsen I've not done a complete review since @seancolsen's already approved the PR.

I've only reviewed the updates based on the comment I had earlier.

File: mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte
Comment (pavish) [reply mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte]: https://github.com/centerofci/mathesar/pull/2489#discussion_r1111313262

File: mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte
Comment (rajatvijay) [reply mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte:38]: @pavish 

>There is an initialInputValue property in ProcessedColumn which should provide the value of the input when the default value is undefined. This eliminates the need for sanitized value.

Could not find this key inside the ProcessedColumn instead using: `getInitialInputValue(column.column);`

>I'm recommending a value & initialValue comparison with stringified values via JSON.stringify since this should provide the value as returned from the server. This is more versatile and works with types like JSON map, JSON list, and array types. The current check would not work for any of them.

There are two reasons for me not taking this approach

**First Reason:** 
When setting a default value for a column with data type number, let's say `value` will be `"1145"` and `initialValue` will be `1145`.
```typescript

$: value = "1145";
$: initialValue = 1145;
$: actionButtionsVisible = JSON.stringify(value) !== JSON.stringify(initialValue);

```

Here actionButtionsVisible will be false should be true since `JSON.stringify("1145") = "\"1145\""` and `JSON.stringify(1145) = "1145"`.

**Second Reason:**
We do not have any such column types as of now that will need JSON map, JSON list, or array types and the purpose of this PR is very different

Comment (pavish) [reply mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte:38]: @rajatvijay 

> Could not find this key inside the ProcessedColumn instead using: getInitialInputValue(column.column);

It's worth adding it to ProcessedColumn. You can refer data explorer which has it.

> We do not have any such column types as of now that will need JSON map, JSON list, or array types and the purpose of this PR is very different

We do have all of these column types implemented, please refer the cell-fabric component. I have disabled it while creating a column for the release. The user is still able to convert a string column to JSON list or map, and the database the user connects to can have these columns.

> Here actionButtionsVisible will be false should be true since JSON.stringify("1145") = "\"1145\"" and JSON.stringify(1145) = "1145".

I'd suggest a type check before this check, instead of `==`. For literals, we could do a `String(value) === String(initialValue)` and we can do a `JSON.stringify` check for objects.

Comment (rajatvijay) [reply mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte]: 385b210dd

Comment (pavish) [reply mathesar_ui/src/systems/table-view/table-inspector/column/SetDefaultValue.svelte]: Looks good

**Pull request merged**
